src/dbnode/storage/series/buffer.go

```diff
@@ -581,38 +580,95 @@ func (b *dbBufferBucket) write(
 	unit xtime.Unit,
 	annotation []byte,
 ) error {
+	datapoint := ts.Datapoint{
+		Timestamp: timestamp,
+		Value:     value,
+	}
+	if b.empty {
+		// No values written yet, write to first buffer
+		return b.writeToEncoderIndex(0, datapoint, unit, annotation)
+	}
+
+	// Find the correct encoder to write to
 	idx := -1
 	for i := range b.encoders {
-		if timestamp.Equal(b.encoders[i].lastWriteAt) {
-			// NB(xichen): We discard datapoints with the same timestamps as the
-			// ones we've already encoded. Immutable/first-write-wins semantics.
-			return nil
+		lastWriteAt := b.encoders[i].lastWriteAt
+
+		if timestamp.Equal(lastWriteAt) {
+			last, err := b.encoders[i].encoder.LastEncoded()
+			if err != nil {
+				return err
+			}
+			if last.Value == value {
+				// No-op since matches the current value
+				// TODO(r): in the future we could return some metadata that
+				// this result was a no-op and hence does not need to be written
+				// to the commit log, otherwise high frequency write volumes
+				// that are using M3DB as a cache-like index of things seen
+				// in a time window will still cause a flood of disk/CPU resource
+				// usage writing values to the commit log, even if the memory
+				// profile is lean as a side effect of this write being a no-op.
+				return nil
+			}
+			continue
 		}
-		if timestamp.After(b.encoders[i].lastWriteAt) {
+
+		if timestamp.After(lastWriteAt) {
 			idx = i
 			break
 		}
 	}
-	if idx == -1 {
-		b.opts.Stats().IncCreatedEncoders()
-		bopts := b.opts.DatabaseBlockOptions()
-		blockSize := b.opts.RetentionOptions().BlockSize()
-		blockAllocSize := bopts.DatabaseBlockAllocSize()
-		encoder := bopts.EncoderPool().Get()
-		encoder.Reset(timestamp.Truncate(blockSize), blockAllocSize)
-		next := inOrderEncoder{encoder: encoder}
-		b.encoders = append(b.encoders, next)
-		idx = len(b.encoders) - 1
+
+	// Upsert/last-write-wins semantics.
```

Comment:
I was skimming through the encoder code and its not clear to me why we need to create a new encoder for this scenario. Does it simplify the iteration logic? it seemed like the encoder could encode multiple different values for the same timestamp

Reply:
It could, it's just way harder and I don't really want to enforce encoders have to deal with this case right now.  We can revisit this later.